src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -1,90 +1,134 @@
+// Their sensitivity score varies based on their output. 
+.decl isTransformationOperation(sql_operation: Operation)
+isTransformationOperation(operation) :- isSqlOperation(operation), 
+      operationHasOperator(operation, operator), 
+      isTransformationOperator(operator). 
+
+.decl isAggregationOperator(op: Operator)
+isAggregationOperator(SQL_SUM_OPERATOR).
+isAggregationOperator(SQL_AVERAGE_OPERATOR).
+isAggregationOperator(SQL_COUNT_OPERATOR).
+isAggregationOperator(SQL_MAX_OPERATOR).
+// Aggregation operations point to where noise needs to be added. 
+.decl isAggregationOperation(sql_operation: Operation)
+isAggregationOperation(operation) :- isSqlOperation(operation), 
+      operationHasOperator(operation, operator),
+      isAggregationOperator(operator).
 
 .decl isSqlValue(value: AccessPath)
 isSqlValue(value) :- isSqlOperation(operation),
       operationHasResult(operation, value).
 
-// keeping track of all initial inputs as they all have sensitivity score 1
-.decl isInitialInput(ingress_point: AccessPath)
+// Keeping track of all initial inputs. 
+.decl isInitialInput (ingress_point: AccessPath)
+isInitialInput(as(ingress_point, AccessPath)) :- isOperation(operation), 
+      operationHasOperator(operation, CORE_INPUT_OPERATOR), 
+      operationHasResult(operation, ingress_point), 
+      operationHasAttribute(operation, [_, $StringAttributePayload(_)]).
 
-// helper functions to flatten input lists
+// Helper functions to flatten input/operand lists. 
 .decl intermediateFlatten(list:OperandList, sublist:OperandList, list_element:AccessPath)
 
-intermediateFlatten(list, sublist, list_element) :- isSqlOperation(operation), 
-      operationHasOperandList(operation, list), 
-      list = [list_element, sublist].
-intermediateFlatten(list, sublist, list_element) :- intermediateFlatten(list, [list_element, sublist], _).
+intermediateFlatten(list, sublist, list_element) :- isOperation(operation), 
+      operationHasOperandList(operation, list),
+      operationHasOperandListSuffix(operation, [list_element, sublist]).
 
 .decl flatten(list:OperandList, list_element:AccessPath)
 flatten(list, list_element) :- intermediateFlatten(list, _, list_element).
 
-// flattening each individual input per value
+// Flattening each individual input per value. 
+
 .decl isSqlInput(operation: Operation, sql_input: AccessPath)
 
 isSqlInput(operation, sql_input) :- isSqlOperation(operation), 
       operationHasOperandList(operation, input_list), 
       flatten(input_list, sql_input). 
 
-// keeping track of sensitivity score per each input at index 
-// used to calculate the maximum input sensitivity 
-// decomposed into base and inductive cases because recursive aggregates are not supported yet 
-// for the compiler case in souffle
-.decl sqlOperationHasSensitivityScoreAtIndex(operation: Operation, sensitivity_score: number, index: number) 
+// Keeping track of sensitivity score per each input at index. 
```

Comment:
I think it could be useful to explain in the comment that `index` is the depth in the graph from the initial inputs (where the initial inputs have index 0).